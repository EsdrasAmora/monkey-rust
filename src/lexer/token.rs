use crate::parser::ast::{BinaryExpression, Expression};
use anyhow::Result;
use serde::{Deserialize, Serialize};
use smol_str::SmolStr;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Token {
    Illegal,
    Identifier(Identifier),
    //Literals
    Int(i64),
    String(SmolStr), // a Cow may would be better here.
    True,
    False,
    Nil,
    //Operators
    Eq,
    NotEq,
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,
    Lt,
    Gt,
    //Delimiters
    Comma,
    Semicolon,
    LParen,
    RParen,
    LBrace,
    RBrace,
    //Keywords
    Function,
    Let,
    If,
    Else,
    Return,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Identifier(SmolStr);

impl Identifier {
    #[inline]
    pub fn new(name: SmolStr) -> Self {
        Self(name)
    }

    #[inline]
    pub fn into_inner(self) -> SmolStr {
        self.0
    }
}

impl TryFrom<Token> for Identifier {
    type Error = Token;

    #[inline]
    fn try_from(token: Token) -> Result<Self, Self::Error> {
        match token {
            Token::Identifier(name) => Ok(name),
            _ => Err(token),
        }
    }
}

impl From<Identifier> for SmolStr {
    #[inline]
    fn from(name: Identifier) -> Self {
        name.into_inner()
    }
}

impl From<SmolStr> for Identifier {
    #[inline]
    fn from(name: SmolStr) -> Self {
        Self::new(name)
    }
}

impl Token {
    //WTF: why I can't wrap the expressions directly using Some()?
    #[inline]
    pub fn binary_expression_type(&self) -> Option<fn(BinaryExpression) -> Expression> {
        Some(match self {
            Token::Plus => Expression::Add,
            Token::Minus => Expression::Sub,
            Token::Slash => Expression::Div,
            Token::Asterisk => Expression::Mul,
            Token::Eq => Expression::Eq,
            Token::NotEq => Expression::NotEq,
            Token::Lt => Expression::Lt,
            Token::Gt => Expression::Gt,
            _ => return None,
        })
    }

    #[inline]
    pub fn precedence(&self) -> u8 {
        match self {
            Token::Eq | Token::NotEq => 2,
            Token::Gt | Token::Lt => 3,
            Token::Plus | Token::Minus => 4,
            Token::Slash | Token::Asterisk => 5,
            Token::LParen => 7,
            _ => 0,
        }
    }

    #[inline]
    pub fn into_identifier(self) -> Option<Identifier> {
        match self {
            Token::Identifier(name) => Some(name),
            _ => None,
        }
    }

    #[inline]
    pub fn is_identifier(&self) -> bool {
        matches!(self, Token::Identifier(_))
    }

    #[inline]
    pub fn is_literal(&self) -> bool {
        matches!(
            self,
            Token::Int(_)
                | Token::String(_)
                | Token::True
                | Token::False
                | Token::Nil
                | Token::Identifier(_)
        )
    }
}
