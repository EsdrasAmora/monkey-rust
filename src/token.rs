use std::fmt::Display;

use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};
use smol_str::SmolStr;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub enum Token {
    Illegal,
    Identifier(Identifier),
    //Literals
    Int(i64),
    String(SmolStr),
    True,
    False,
    Nil,
    Dot,
    //Operators
    Eq,
    NotEq,
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,
    Lt,
    Lte,
    Gt,
    Gte,
    //Delimiters
    Comma,
    Colon,
    Semicolon,
    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,
    //Keywords
    Function,
    Let,
    If,
    Else,
    Return,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Identifier(SmolStr);

impl Identifier {
    #[inline]
    pub fn new(name: SmolStr) -> Self {
        Self(name)
    }

    #[inline]
    pub fn inner(&self) -> SmolStr {
        self.0.clone()
    }
}

impl Display for Identifier {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.0.fmt(f)?;
        Ok(())
    }
}

impl TryFrom<Token> for Identifier {
    type Error = anyhow::Error;

    #[inline]
    fn try_from(token: Token) -> Result<Self, Self::Error> {
        match token {
            Token::Identifier(name) => Ok(name),
            _ => Err(anyhow!("Expected identifier but found: {:?}", token)),
        }
    }
}

impl Token {
    #[inline]
    pub fn precedence(&self) -> u8 {
        match self {
            Token::Eq | Token::NotEq => 2,
            Token::Gt | Token::Gte | Token::Lt | Token::Lte => 3,
            Token::Plus | Token::Minus => 4,
            Token::Slash | Token::Asterisk => 5,
            Token::LParen => 7,
            Token::LBracket => 8,
            _ => 0,
        }
    }

    #[inline]
    pub fn is_literal(&self) -> bool {
        matches!(
            self,
            Token::Int(_)
                | Token::String(_)
                | Token::True
                | Token::False
                | Token::Nil
                | Token::Identifier(_)
        )
    }
}
