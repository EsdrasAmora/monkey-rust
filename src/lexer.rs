use crate::token::{Identifier, Token};
use smol_str::SmolStr;
use std::iter::{self, Peekable};

#[derive(Debug)]
pub struct Lexer {
    pub tokens: Vec<Token>,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        let mut chars = input.trim().chars().filter(|x| x.is_ascii()).peekable();
        let mut tokens = Vec::with_capacity(32);

        while let Some(char) = chars.next() {
            if char.is_whitespace() {
                continue;
            }
            let token = Lexer::new_helper(char, &mut chars);
            tokens.push(token);
        }

        Lexer { tokens }
    }

    fn new_helper(char: char, chars: &mut Peekable<impl Iterator<Item = char>>) -> Token {
        match char {
            '+' => Token::Plus,
            ',' => Token::Comma,
            ';' => Token::Semicolon,
            '(' => Token::LParen,
            ')' => Token::RParen,
            '{' => Token::LBrace,
            '}' => Token::RBrace,
            '[' => Token::LBracket,
            ']' => Token::RBracket,
            '-' => Token::Minus,
            '*' => Token::Asterisk,
            '/' => Token::Slash,
            ':' => Token::Colon,
            '<' => chars.next_if_eq(&'=').map_or(Token::Lt, |_| Token::Lte),
            '>' => chars.next_if_eq(&'=').map_or(Token::Gt, |_| Token::Gte),
            '=' => chars.next_if_eq(&'=').map_or(Token::Assign, |_| Token::Eq),
            '!' => chars.next_if_eq(&'=').map_or(Token::Bang, |_| Token::NotEq),
            '"' => {
                let string: SmolStr = iter::from_fn(|| chars.next_if(|x| *x != '"')).collect();
                chars.next();
                Token::String(string)
            }
            _ if char.is_ascii_alphabetic() || char == '_' => {
                let keyword: SmolStr = iter::once(char)
                    .chain(iter::from_fn(|| {
                        chars.next_if(|x| x.is_ascii_alphanumeric() || *x == '_')
                    }))
                    .collect();

                match keyword.as_str() {
                    "fn" => Token::Function,
                    "let" => Token::Let,
                    "true" => Token::True,
                    "false" => Token::False,
                    "if" => Token::If,
                    "else" => Token::Else,
                    "return" => Token::Return,
                    "nil" => Token::Nil,
                    _ => Token::Identifier(Identifier::new(keyword)),
                }
            }
            _ if char.is_ascii_digit() => iter::once(char)
                .chain(iter::from_fn(|| chars.next_if(|x| x.is_ascii_digit())))
                .collect::<SmolStr>()
                .parse()
                .map_or(Token::Illegal, Token::Int),
            _ => Token::Illegal,
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;

    #[test]
    fn tokenize_sucessfully() {
        let input = r#" 
        let five = 5;
        let ten = 10;

        let add = fn(x, y) {
            x + y;
        };

        let result = add(five, ten);
        !-/*5
        5 < 10 > 5

        if (5 < 10) {
            return true;
        } else {
            return false;
        }

        10 == 10
        10 != 9
        "foobar"
        "foo bar"
        [1, 2]
        {"foo": "bar"}
        "#;

        let result = Lexer::new(input);
        assert_yaml_snapshot!(result.tokens);
    }
}
