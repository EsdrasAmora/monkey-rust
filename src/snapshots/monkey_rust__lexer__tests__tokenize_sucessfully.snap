---
source: src/lexer.rs
expression: result.tokens
---
- Let
- Identifier: five
- Assign
- Int: 5
- Semicolon
- Let
- Identifier: ten
- Assign
- Int: 10
- Semicolon
- Let
- Identifier: add
- Assign
- Function
- LParen
- Identifier: x
- Comma
- Identifier: y
- RParen
- LBrace
- Identifier: x
- Plus
- Identifier: y
- Semicolon
- RBrace
- Semicolon
- Let
- Identifier: result
- Assign
- Identifier: add
- LParen
- Identifier: five
- Comma
- Identifier: ten
- RParen
- Semicolon
- Bang
- Minus
- Slash
- Asterisk
- Int: 5
- Int: 5
- Lt
- Int: 10
- Gt
- Int: 5
- If
- LParen
- Int: 5
- Lt
- Int: 10
- RParen
- LBrace
- Return
- "True"
- Semicolon
- RBrace
- Else
- LBrace
- Return
- "False"
- Semicolon
- RBrace
- Int: 10
- Eq
- Int: 10
- Int: 10
- NotEq
- Int: 9
- String: foobar
- String: foo bar
- LBracket
- Int: 1
- Comma
- Int: 2
- RBracket
- LBrace
- String: foo
- Colon
- String: bar
- RBrace

